### PR TITLE
#958 unblocking tedge mapper start up

### PR DESCRIPTION
**Pull request by initard**
## Proposed changes

unblocked tedge mapper startup. http_proxy was causing the mapper to remain in a while loop trying to fetch an internal device id from cumulocity. this has now been taken out of tedge mapper startup.  

the struct in question, C8yEndPoint, now has an uninitialised version (C8yEndPointUninitialised) and the either crate is used to change the state from uninitialised to initialised. 

the full proposal is covered [here](https://github.com/thin-edge/thin-edge.io/discussions/915#discussioncomment-2630789)

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [x] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
## Paste Link to the issue
- Closes to #958 

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)

### CODE REVIEW

File: crates/core/c8y_api/Cargo.toml [hunk 1]
```diff
@@ -14,6 +14,7 @@ c8y_smartrest = { path = "../c8y_smartrest" }
 clock = { path = "../../common/clock" }
 csv = "1.1"
 download = { path = "../../common/download" }
+either = "1.6"
```
Comment (makr11st) [inline crates/core/c8y_api/Cargo.toml:17]: I'm a little bit confused with the use of `either` crate here? Why not go with a straight forward enum of structs?
With the crate's `left` and `right` variants are vague and make reading the code a bit of a hassle to find out which one variant are we using here.

File: crates/core/c8y_api/src/http_proxy.rs [hunk 2]
```diff
@@ -313,30 +331,41 @@ impl JwtAuthHttpProxy {
...
+            info!("Initialisation");
+
+            let uninitialised = self.end_point.as_ref().unwrap_left().clone();
+            let c8y_end_point = match self.try_get_internal_id().await {
+                Ok(internal_id) => Ok(C8yEndPoint::new(&uninitialised, internal_id)),
+                Err(_error) => Err(SMCumulocityMapperError::FailedToRetrieveC8yInternalId),
+            }?;
+            self.end_point = either::Either::Right(c8y_end_point);
+            info!("Initialisation done.");
         }
```
Comment (makr11st) [inline crates/core/c8y_api/src/http_proxy.rs:357]: Wh ynot just go with `if let` construct here (sorry for pseudo code):
```suggestion
        if let Uninitialized = self.end_point {
            info!("Initialisation");

            let uninitialised = self.end_point.as_ref().unwrap_left().clone();
            let c8y_end_point = match self.try_get_internal_id().await {
                Ok(internal_id) => Ok(C8yEndPoint::new(&uninitialised, internal_id)),
                Err(_error) => Err(SMCumulocityMapperError::FailedToRetrieveC8yInternalId),
            }?;
            self.end_point = either::Either::Right(c8y_end_point);
            info!("Initialisation done.");
        }
```

**Review by makr11st — COMMENTED** [review]
I'm surprised that we are discarding the message when initialization fails, even though it is logged the event / software list request will not be triggered again.

I believe that in this case we should either retry at least or queue the messages to be sent whenever we can initialize properly.

It was pointed to me by Alex that we foresee this initialisation to happen almost immediately after the mapper starts, but it would not be a case as it will only be initialised when an event requiring HTTP will occur. I believe the right way to go about this is to have the init run as a task and be triggered at the start, but also we would need a caching for messages which cant be send while uninitialised.

File: crates/core/c8y_api/src/http_proxy.rs (hunk 2 above)
Comment (didier-wenzek) [reply crates/core/c8y_api/src/http_proxy.rs:357]: Indeed, this will avoid the following `unwrap_left()`.

File: crates/core/c8y_api/Cargo.toml (hunk 1 above)
Comment (didier-wenzek) [reply crates/core/c8y_api/Cargo.toml:17]: Both approaches - using the `either` crate or having some specific implementation, make sense. I agree that what is `Left` or `Right` cannot be guessed without a check to the definition of the `JwtAuthHttpProxy::end_point`. So it might be a good idea switch to a specific implementation - unless one find ourselves re-implementing helper functions . @initard can you check that we mostly use pattern matching to manipulate end-points?

Comment (initard) [reply crates/core/c8y_api/Cargo.toml:17]: I believe it is the case indeed. We should move it to a specific implementation.

**Comment by didier-wenzek** [comment]
> I'm surprised that we are discarding the message when initialization fails, even though it is logged the event / software list request will not be triggered again.

Indeed, the fix proposed here unblocks the mapper - that can process local messages even if there is no cloud connections, but makes another issue more likely - if there is no cloud connections when the mapper start then the current software list is not sent to the cloud.

> I believe that in this case we should either retry at least or queue the messages to be sent whenever we can initialize properly.

A retry would just push the issue a step further. A queue of messages to be sent would a better fix. If we focus on the issue that the current software list is not sent to the cloud when there is no cloud connections on start, I think we can find some not too complex workaround. But I'm afraid this workaround will break sooner or later.

> It was pointed to me by Alex that we foresee this initialisation to happen almost immediately after the mapper starts, but it would not be a case as it will only be initialised when an event requiring HTTP will occur. I believe the right way to go about this is to have the init run as a task and be triggered at the start, but also we would need a caching for messages which cant be send while uninitialised.

Fixing the issue for all the events requiring HTTP requires much more that a queue of messages to be retried. We need to rethink the boundaries between the mapper and the HTTP proxy. Indeed, a method like `C8YHttpProxy::send_event` or  `C8YHttpProxy::upload_log_binary` is *partial* and focused on the HTTP part of the request. In the current code base, when these HTTP related methods return then the mapper needs to finalize the request addressing the MQTT part of the Cumulocity protocol. So, if we let this separation of work unchanged, not only one needs a queue of messages to be processed by the http proxy when the network will be up, one also needs a queue of messages to be further processed by the mapper.

I think the mistake is to cut the responsibilities between the mapper and the http proxy around the transport protocol while Cumulocity uses a mix of HTTP and MQTT.

In conclusion, I think that what to be done is a bit involved:
* Reshape the HTTPProxy interface so a request like `upload_log_binary` is processed in its entirety (getting the JWT token from MQTT, uploading the content to cloud using HTTP, fetching the log event url *and* sending this url to the cloud via MQTT). 
* The mapper can then fully delegate these requests to the proxy without being blocked.

**Comment by didier-wenzek** [comment]
Closing this abandoned PR
